### PR TITLE
Create task list REST API endpoint

### LINKED ISSUE #7365: Create tasks REST API endpoint
Tasks should be retrievable via a REST endpoint.  There is already a controller in place for onboarding tasks, so this will issue will involve adding an endpoint.

**Endpoint**
`GET` - `/wc-admin/onboarding/tasks`

**Schema**

Since we have multiple task lists, this endpoint will ideally return the tasks lists as an array with tasks already organized under each list.

```php
array(
    'id'         => <string>,
    'isComplete' => <bool>,
    'isHidden'   => <bool>,
    'title'      => <string>,
    'tasks'      => array(
        array(
            'id'          => <string>,
            'title'       => <string>,
            'content'     => <string>,
            'actionLabel' => <string>,
            'actionUrl'   => <string>,
            'isComplete'  => <bool>,
            'isVisible'   => <bool>,
            'time'        => <string>,
        )
    ),
)
```

https://github.com/woocommerce/woocommerce-admin/pull/7354 contains code for this endpoint, but will need additional tests and schema.

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #7365 

Adds the REST endpoint needed for the task list and migrates all existing tasks to the server.

### Screenshots

Screen Shot 2021-08-16 at 4 34 48 PM

### Detailed test instructions:

1. Create a `GET` request to `wp-json/wc-admin/onboarding/tasks`
2. Note the tasks in the response
3. Make sure the `isComplete` and `isVisible` properties are the correct values for each task:

| Task  | Visible  |  Complete |
|---|---|---|
| Store Details  | Always  | Complete store profiler  |
|  Paid extensions | Select paid products during "product types" of store profiler  | Install all purchase-able plugins |
| Products  | Always  | Add at least one product  |
| WooCommerce Payments  | Select WC Pay for install during business extensions step in profiler  |  Install and connect WC Pay  |
| Payments  |  Don't select WC Pay during business extensions step in profiler and don't install WooCommerce Payments | Install at least one payment gateway  |
| Tax  | Always  | Connect taxes through WCS or add at least one tax code  |
| Shipping  | Create at least one physical product  |  Create at least one shipping zone |
|  Marketing | Always (assuming remote free extensions feature is enabled  |  Install at least one marketing extension from the list |
|  Appearance |  Always | `woocommerce_task_list_appearance_complete` is truthy; walk through the appearance task steps  |

### CODE REVIEW

File: src/API/OnboardingTasks.php [hunk 1]
```diff
@@ -610,4 +641,230 @@ public function get_status() {
...
+							'woocommerce-admin'
+						),
+						'actionLabel' => __( "Let's go", 'woocommerce-admin' ),
+						'actionUrl'   => '/setup-wizard',
+						'isComplete'  => isset( $profiler_data['completed'] ) && true === $profiler_data['completed'],
+						'isVisible'   => true,
+						'time'        => __( '4 minutes', 'woocommerce-admin' ),
+					),
+					array(
+						'id'            => 'purchase',
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:724]: This task is currently rendering as "Add physical to my store" for me; I see you have a couple todo's around this, but thought I'd mention. I'm not precisely sure what this task is intended for but it seems that it'll be handled in another PR?

File: src/API/OnboardingTasks.php [hunk 2]
```diff
@@ -610,4 +641,230 @@ public function get_status() {
...
+						'isVisible'  => Features::is_enabled( 'payment-gateway-suggestions' ) &&
+							(
+								! in_array( 'woocommerce-payments', $business_extensions, true ) ||
+								! in_array( 'woocommerce-payments', $installed_plugins, true ) ||
+								! in_array( WC()->countries->get_base_country(), OnboardingTasksFeature::get_woocommerce_payments_supported_countries(), true )
+							),
+						'time'       => __( '2 minutes', 'woocommerce-admin' ),
+					),
+					array(
+						'id'          => 'tax',
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:798]: Since this list is extendable, I wonder if using more unique IDs would be worth-while to avoid collisions down the road?

File: src/API/OnboardingTasks.php [hunk 3]
```diff
@@ -610,4 +641,230 @@ public function get_status() {
 
 		return rest_ensure_response( $status );
 	}
+
+	/**
+	 * Get the onboarding tasks.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function get_tasks() {
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:650]: This function will get cluttered over time, with all of the queries/logic regarding every task living here. Not a necessary optimization yet, but we may want to modularize the logic per-task at some point as the list grows more complex.

**Review by joelclimbsthings — APPROVED** [review]
Absolutely well done @joshuatf , working great for me. I left a few comments, but nothing mandatory. Pre-approved 🚢

File: src/API/OnboardingTasks.php (hunk 1 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:737]: I think it's best to go ahead and handle here.  Pushed some updates in https://github.com/woocommerce/woocommerce-admin/pull/7512/commits/859fe940ebfc05d731fa774742055a6eead74ba8.

File: src/API/OnboardingTasks.php (hunk 2 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:812]: I'm not opposed to this, but since these IDs already existed, we would be breaking any backwards compatibility with filters being used.  But if we think they're not being used at all we can forego cc @octaedro

File: src/API/OnboardingTasks.php (hunk 3 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:650]: Completely agree.  This along with most of the onboarding feature classes could benefit from being split up into smaller, task-related classes.

I think I may try and tackle this during next cooldown.

**Comment by joshuatf** [comment]
Hey @joelclimbsthings, I added those text changes for the purchaseable products in https://github.com/woocommerce/woocommerce-admin/pull/7512/commits/859fe940ebfc05d731fa774742055a6eead74ba8 - could you give this another look before merge?

File: src/API/OnboardingTasks.php [hunk 4]
```diff
@@ -610,4 +641,244 @@ public function get_status() {
...
+							: __(
+								'Add paid extensions to my store',
+								'woocommerce-admin'
+							),
+						'content'       => count( $remaining_products ) === 1
+							? $purchaseable_products[0]['description']
+							: sprintf(
+								/* translators: %1$s: list of product names comma separated, %2%s the last product name */
+								__(
+									'Good choice! You chose to add %1$s and %2$s to your store.',
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:756]: The output for this line on my local is "...You chose to add  and  to your store."

image

File: src/API/OnboardingTasks.php [hunk 5]
```diff
@@ -610,4 +641,244 @@ public function get_status() {
...
+
+	/**
+	 * Get the onboarding tasks.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function get_tasks() {
+		$profiler_data         = get_option( Onboarding::PROFILE_DATA_OPTION, array() );
+		$installed_plugins     = PluginsHelper::get_installed_plugin_slugs();
+		$product_types         = $profiler_data['product_types'];
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:653]: I just noticed that a fresh site may throw an error on this line since the default for `PROFILE_DATA_OPTION` is an empty array:

image

**Review by joelclimbsthings — CHANGES_REQUESTED** [review]
Thanks for the changes @joshuatf . I added one comment regarding the text output, and also came across an error state that could be encountered on a fresh site. See my comments above.

File: src/API/OnboardingTasks.php (hunk 4 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:756]: I think `isVisible` for this would be false, so we can either filter these out so they are not shown or set the `content` to `null`.  But I'd be partial to the former if we want to hide this.

File: src/API/OnboardingTasks.php (hunk 5 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php]: Good catch!  Updated in https://github.com/woocommerce/woocommerce-admin/pull/7512/commits/c181247c20465b4fe2b5199d7af9c27f0571c0e4

**Review by joelclimbsthings — APPROVED** [review]
Awesome work @joshuatf . I hereby restore my approval ✅ :shipit:
